docs/multi-currency/multi-currency.md

```diff
@@ -0,0 +1,134 @@
+
+To enable the forging of currency value, the paper [2] envisages a new type of transaction called `CurrencyTx` (see [2, Def. 2]). `CurrencyTx` creates a new *known currency*, by registering the currency's name as a unique (across the ledger) string, and associating it with a `Script` representing the monetary policy for that currency. 
+
+Adding the `CurrencyTx` type results in additional work for core nodes: They need to keep track of the name and monetary policy of every currency that has ever been created. So `CurrencyTx` adds a new kind of ledger-wide, global state (in addition to the UTXO set), in form of the registry of currencies and their policies. Unlike the UTXO set this new global state can only ever grow larger, because there is no way to destroy a currency. While at the moment neither the cost model of regular UTXO-with-script transactions nor that of `CurrencyTx` transactions has been developed, it seems likely that the cost of a `CurrencyTx` is going to be higher than the effect an added `PayToScript` output has on the cost of a regular transaction. 
+
+This makes multi-currency based on the `CurrencyTx` proposal potentially unsuitable for applications that require lightweight currencies, in particular the implementation of [Non-Fungible Tokens](https://eips.ethereum.org/EIPS/eip-721) (NFTs).
+
+## Proposed solution
+
+We propose a new extension to the UTXO model with scripts, an extension that has the same effects as adding a `CurrencyTx` transaction type (namely, the ability to forge value of user-defined currencies) but without the added overhead of a currency registry.
+
+The proposal consists of two changes.
+
+1. Instead of using a `String` value to identify a currency, we use the `ScriptHash` type (a ByteString)
+2. We replace the **(forging)** rule above with the following rule: **(forging2)** A transaction with a non-zero `txForge` field is only valid if for every key `h` in `txForge`, the transaction's set of inputs `txIn` contains a `TxIn` that spends a `TxOut` whose address is `h`.
+
+The **(forging2)** rule, together with the **(legitimacy)** rule, ensures that the script `H` (with hash `h`) is run whenever new value of the currency `h` is forged. Because `H` is provided with the `txForge` field of the spending transaction, via the `PendingTx` type, it can block or authorise any amount of value that is created of its currency. There is no need for a registry of currencies because the thing that identifies a currency is (the hash of) its monetary policy.
+
+With this proposal a custom currency is no different from any other smart contract, and currencies don't require a separate cost model.
+
+We can write monetary policy as a state machine that keeps track of the current supply and forges more value when necessary.
```

Comment:
This requires us to be able to write secure state machines, which we still don't quite know how to do, right?

Reply:
We know how to do it assuming we can deserialise data scripts from PLC back to Haskell. Multi-currency based NFTs work quite well with that because the entire state of the contract can be in one data script and you can check whether the appropriate permissions are there by looking at the NFTs that are spent & produced by the transaction

Reply:
Sure - but if someone asks we still don't know how we're going to do that!

Reply:
True!